src/Servers/Kestrel/Core/src/Internal/Infrastructure/PipeWriterHelpers/ConcurrentPipeWriter.cs

```diff
@@ -0,0 +1,382 @@
+            }
+        }
+
+        private async ValueTask<FlushResult> FlushAsyncAwaited(ValueTask<FlushResult> flushTask, CancellationToken cancellationToken)
+        {
+            try
+            {
+                // This while (true) does look scary, but the real continuation condition is at the start of the loop
+                // after the await, so the _sync lock can be acquired.
+                while (true)
+                {
+                    var flushResult = await flushTask;
+
+                    lock (_sync)
+                    {
+                        if (_bytesBuffered == 0)
+                        {
+                            CompleteFlushUnsynchronized(flushResult, null);
+                            return flushResult;
+                        }
+
+                        CopyAndReturnSegmentsUnsynchronized();
+
+                        flushTask = _innerPipeWriter.FlushAsync(cancellationToken);
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                lock (_sync)
+                {
+                    CompleteFlushUnsynchronized(default, ex);
+
+                    // Ensure the exception still gets observed by the original caller who started the flush loop
+                    // And is observing this method directly instead of through the TCS.
+                    throw;
+                }
+            }
+        }
+
+        // This is not exposed to end users. Throw so we find out if we ever start calling this.
+        public override void OnReaderCompleted(Action<Exception, object> callback, object state)
+        {
+            throw new NotImplementedException();
+        }
+
+        public void Abort()
+        {
+            lock (_sync)
+            {
+                _aborted = true;
+
+                // If we're flushing, the cleanup will happen after the flush.
+                if (_currentFlushTcs == null)
+                {
+                    CleanupUnsynchronized();
+                }
+            }
+        }
+
+        private void CleanupUnsynchronized()
+        {
+            BufferSegment segment = _head;
+            while (segment != null)
+            {
+                BufferSegment returnSegment = segment;
+                segment = segment.NextSegment;
+                returnSegment.ResetMemory();
+            }
+
+            _head = null;
+            _tail = null;
+            _tailMemory = null;
+        }
+
+        private void CopyAndReturnSegmentsUnsynchronized()
+        {
+            // Update any buffered data
+            _tail.End += _tailBytesBuffered;
+            _tailBytesBuffered = 0;
+
+            var segment = _head;
+            var returnSegment = default(BufferSegment);
+
+            while (segment != null)
+            {
+                // Make sure we don't return the last segment in the linked list yet.
+                if (returnSegment != null)
```

Comment:
Feels weird to do this at the beginning of the loop. Why not have it at the end and check segment.NextSegment? That would avoid having the return segment entirely.

Reply:
I don't want to decide whether or not to return the tail segment until the end of the loop. This logic ensure that returnSegment will never be the tail segment. There are a lot of ways to write this, but I don't think the other ways are any clearer if you want to have special logic determining whether or not to return the tail segment.

Reply:
Nvm. It does look a bit cleaner the way you suggested.